File: 42.trapping-rain-water.ts
/*
 * @lc app=leetcode id=42 lang=typescript
 *
 * [42] Trapping Rain Water
 */

// @lc code=start
function trap(height: number[]): number {
  let l = 0
  let r = height.length - 1
  let lMax = height[l]
  let rMax = height[r]

  let res = 0

  while (l < r) {
    if (lMax < rMax) {
      l++
      lMax = Math.max(lMax, height[l])
      res += lMax - height[l]
    } else {
      r--
      rMax = Math.max(rMax, height[r])
      res += rMax - height[r]
    }
  }

  return res
};
// @lc code=end

File: python/42.trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        l = 0
        r = n-1
        l_max = 0
        r_max = 0
        total_water = 0
        while l <= r:
            if height[l] <= height[r]:
                if height[l] >= l_max:
                    l_max = height[l]
                else:
                    total_water += l_max - height[l]
                l += 1
            
            else:
                if height[r] >= r_max:
                    r_max = height[r]
                else:
                    total_water += r_max - height[r]
                r -= 1
        return total_water

        